plugins/auth-node/src/DefaultIdentityClient.ts

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright 2020 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { PluginEndpointDiscovery } from '@backstage/backend-common';
+import { AuthenticationError } from '@backstage/errors';
+import {
+  createRemoteJWKSet,
+  decodeJwt,
+  decodeProtectedHeader,
+  FlattenedJWSInput,
+  JWSHeaderParameters,
+  jwtVerify,
+} from 'jose';
+import { GetKeyFunction } from 'jose/dist/types/types';
+
+import { BackstageIdentityResponse } from './types';
+import { getBearerTokenFromAuthorizationHeader, IdentityApi } from '.';
+
+const CLOCK_MARGIN_S = 10;
+
+/**
+ * An identity client options object which allows extra configurations
+ *
+ * @experimental This is not a stable API yet
+ * @public
+ */
+export type IdentityClientOptions = {
+  discovery: PluginEndpointDiscovery;
+  issuer?: string;
+
+  /** JWS "alg" (Algorithm) Header Parameter values. Defaults to an array containing just ES256.
+   * More info on supported algorithms: https://github.com/panva/jose */
+  algorithms?: string[];
+};
+
+/**
+ * An identity client to interact with auth-backend and authenticate Backstage
+ * tokens
+ *
+ * @experimental This is not a stable API yet
+ * @public
+ */
+export class DefaultIdentityClient implements IdentityApi {
+  private readonly discovery: PluginEndpointDiscovery;
+  private readonly issuer?: string;
+  private readonly algorithms?: string[];
+  private keyStore?: GetKeyFunction<JWSHeaderParameters, FlattenedJWSInput>;
+  private keyStoreUpdated: number = 0;
+
+  /**
+   * Create a new {@link DefaultIdentityClient} instance.
+   */
+  static create(options: IdentityClientOptions): DefaultIdentityClient {
+    return new DefaultIdentityClient(options);
+  }
+
+  private constructor(options: IdentityClientOptions) {
+    this.discovery = options.discovery;
+    this.issuer = options.issuer;
+    this.algorithms = options.hasOwnProperty('algorithms')
+      ? options.algorithms
+      : ['ES256'];
+  }
+
+  async getIdentity(req: Request) {
+    try {
+      return await this.authenticate(
+        getBearerTokenFromAuthorizationHeader(req.headers.get('authorization')),
+      );
+    } catch (e) {
+      return undefined;
```

Comment:
I am undecided about this returning undefined. In a way the user has not been authenticated. However I think I feel like its not up to this function to raise an error if the user cannot be authenticated. That might be better done by another middleware up the stack. e.g.

(req, res) => {
  if (!(await identity.getIdentity(req)) {
    res.send(401)
  }
}

Reply:
Done.